keg_relocate: only replace matches at the start of a path

Rylan12 opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

See #11302

This PR modifies the matching used in `keg_relocate` so that prefix/cellar/library/repository paths are only replaced with a placeholder if they appear at the start of the path. For example, this:

```sh
cat /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/objc-run.rb
cat /foo/bar/usr/local/baz.txt
```

Becomes this:

```sh
cat @@HOMEBREW_LIBRARY@@/Taps/homebrew/homebrew-core/Formula/objc-run.rb
cat /foo/bar/usr/local/baz.txt
```

I've tested this locally with the `objc-run` formula (see https://github.com/Homebrew/homebrew-core/pull/76394) and it appears to work:

```console
$ diffoscope existing-objc-run--1.4.big_sure.bottle.1.tar.gz new-objc-run--1.4.big_sur.bottle.2.tar.gz
--- existing-objc-run--1.4.big_sure.bottle.1.tar.gz
+++ new-objc-run--1.4.big_sur.bottle.2.tar.gz
│   --- existing-objc-run--1.4.big_sure.bottle.1.tar
├── +++ new-objc-run--1.4.big_sur.bottle.2.tar
│ ├── objc-run/1.4/bin/objc-run
│ │ @@ -103,15 +103,15 @@
│ │
│ │     )
│ │     xcodebuild -workspace "$projectDirPath/clitest.xcworkspace/" -scheme clitest -configuration Release DEPLOYMENT_LOCATION=YES DSTROOT=build DWARF_DSYM_FOLDER_PATH=build
│ │     xcodebuildExitCode=$?
│ │
│ │     # on clang success, run compiled application and remove it
│ │     if [[ $xcodebuildExitCode -eq 0 ]]; then
│ │ -           cp "$projectDirPath/build@@HOMEBREW_PREFIX@@/bin/clitest" $target
│ │ +           cp "$projectDirPath/build/usr/local/bin/clitest" $target
│ │             trap '{ rm -f "$target" ; exit 1; }' INT
│ │             "$target" "$@"
│ │             trap '' INT
│ │             rm -f "$target"
│ │     else
│ │             echo "objc-run: xcodebuild returned with error" 1>&2
│ │             exit 1
```

The `@@HOMEBREW_PREFIX@@` placeholder is no longer used in the new bottle.

---

Some other notes:

Right now I used `[a-zA-Z0-9]` as the pattern for characters _not_ to match before the given path. This means that anything like `a/usr/local`, `A/usr/local`, `0/usr/local` will not cause the placeholders to be used. If there's a better pattern to use I'm open to it. I don't think these should cause any issues, though, as I can't think of a situation in which the text would match that pattern but still need a placeholder.

Also, I haven't really tested this througoughly and I'm not totally sure what the best way to do so it. I've tested on `objc-run`, but that's it. Is there anywhere in `brew tests` that tests the placeholder substitution? I wasn't able to find one but if there is, I'd like to add a test to ensure that these patterns aren't matched.

carlocab commented on Library/Homebrew/keg_relocate.rb:
```diff
@@ -81,7 +81,11 @@ def replace_text_in_files(relocation, files: nil)
-      changed = s.gsub!(Regexp.union(replacements.keys.sort_by(&:length).reverse), replacements)
+
+      replacement_regexes = replacements.keys.sort_by(&:length).reverse.map do |old_path|
+        /(?<=[^a-zA-Z0-9])#{old_path}/
```
```suggestion
        /(?<![a-zA-Z0-9])#{old_path}/
```
?

carlocab reviewed (approved):
LGTM. I was initially concerned by formulae that record paths in text files with something silly like
```
PATH/usr/local/bin
```
but it seems sufficiently unlikely. (Plus is probably also software that deserves to get broken.)

Rylan12 commented:
Also, CC @sjackman in case you want to take a look (as this deals with relocation)

MikeMcQuaid commented on Library/Homebrew/keg_relocate.rb:
```diff
@@ -81,7 +81,11 @@ def replace_text_in_files(relocation, files: nil)
-      changed = s.gsub!(Regexp.union(replacements.keys.sort_by(&:length).reverse), replacements)
+
+      replacement_regexes = replacements.keys.sort_by(&:length).reverse.map do |old_path|
+        /(?<![a-zA-Z0-9])#{old_path}/
```
Wondering if there's any other characters we need to cover e.g. `._-` or similar? Maybe not.

Rylan12 replied:
Yeah, I'm not sure. There are a whole bunch of things that are possible. It's probably just a game of "choose which are most-often used and, for any others that show up, either update this, use `inreplace`,  or use `pour_bottle?`.

- `./usr/local`: Probably not because relative paths are likely not actually referring to the prefix
- `_/usr/local`: Idk where this would show up so it's hard to say. It doesn't really seem like it would be referring to a path, though, so I'd say don't include unless it's a problem?
- `-/usr/local`: Again, not sure. Is it possible that it's trying to subtract the `/usr/local` prefix from another path? Might be worth including, then.

I'm really just not sure what kinds of things are used so unless anyone has anything specific they _know_ is being missed we may just want to move forward and handle new cases that come up?

Bo98 replied:
I'd love to collect data on this at some point but that's not trivial to do.

MikeMcQuaid replied:
The more I think about it: I think it's fine as-is.

> I'm really just not sure what kinds of things are used so unless anyone has anything specific they _know_ is being missed we may just want to move forward and handle new cases that come up?

Agreed!

MikeMcQuaid reviewed (approved):
I'm happy when you are!

Bo98 commented:
> I was initially concerned by formulae that record paths in text files with something silly like

If that happens the fix should be in the formula itself. Automatic relocation should really never have "false replacements".

Bo98 commented:
Thinking about this more: what about `keg_contain?` in `dev-cmd/bottle`? What happens to the `cellar` value for something containing `build/usr/local` like that one formula did?

MikeMcQuaid reviewed (changes requested):
Pretty sure `each_unique_file_matching` will need updated here too (thanks @Bo98 for noticing)

Rylan12 commented:
I noticed that, currently, `each_unique_file_matching` uses `fgrep` which, according to the manpage:

```
fgrep is quicker than both grep and egrep, but can only handle fixed patterns
     (i.e. it does not interpret regular expressions).
```

So, it looks like we'll need to use `egrep` instead. Is that okay or is there another reason to stick to `fgrep`. Or if there a different way this should be handled?

carlocab commented:
I suspect we were using `fgrep` for the performance benefit.

Rylan12 commented:
Interestingly:

```console
❯ hyperfine -i 'fgrep -lrO "/usr/local" mingw-w64' 'egrep -lrO "/usr/local" mingw-w64' --warmup 2
Benchmark #1: fgrep -lrO "/usr/local" mingw-w64
  Time (mean ± σ):      6.800 s ±  0.970 s    [User: 5.799 s, System: 0.554 s]
  Range (min … max):    6.012 s …  8.661 s    10 runs

  Warning: Ignoring non-zero exit code.

Benchmark #2: egrep -lrO "/usr/local" mingw-w64
  Time (mean ± σ):      6.142 s ±  0.218 s    [User: 5.585 s, System: 0.442 s]
  Range (min … max):    5.990 s …  6.713 s    10 runs

  Warning: Ignoring non-zero exit code.

Summary
  'egrep -lrO "/usr/local" mingw-w64' ran
    1.11 ± 0.16 times faster than 'fgrep -lrO "/usr/local" mingw-w64'
```

Although, it looks like neither the default `grep` or `egrep` support negative look behinds which is what we're using in the other part of this PR.

I just did a quick look at `mingw-w64` (the largest formula I currently have installed) and noticed this:

```console
$ fgrep -lrO "/usr/local" mingw-w64 | wc
fgrep: mingw-w64/8.0.0_4/toolchain-i686/mingw: Is a directory
fgrep: mingw-w64/8.0.0_4/toolchain-x86_64/mingw: Is a directory
     363     363   23947

$ rg -lLP '/usr/local' mingw-w64 | wc
     121     121    8492
```

There seem to be a pretty different number of results obtained so I wonder if I don't have the flags set right...

But much much faster:
```console
$ hyperfine -i 'fgrep -lrO "/usr/local" mingw-w64' 'rg -lLP "(?<![a-zA-Z0-9])/usr/local" mingw-w64' --warmup 2
Benchmark #1: fgrep -lrO "/usr/local" mingw-w64
  Time (mean ± σ):      6.221 s ±  0.257 s    [User: 5.621 s, System: 0.457 s]
  Range (min … max):    6.039 s …  6.826 s    10 runs

  Warning: Ignoring non-zero exit code.

Benchmark #2: rg -lLP "(?<![a-zA-Z0-9])/usr/local" mingw-w64
  Time (mean ± σ):     767.4 ms ± 157.9 ms    [User: 1.484 s, System: 0.710 s]
  Range (min … max):   661.0 ms … 1163.0 ms    10 runs

Summary
  'rg -lLP "(?<![a-zA-Z0-9])/usr/local" mingw-w64' ran
    8.11 ± 1.70 times faster than 'fgrep -lrO "/usr/local" mingw-w64'
```

I'll plan on looking into this a little more this weekend if I don't get to it later tonight.

Rylan12 commented:
I think the change doesn't need to be made in `each_unique_file_matching` as that just needs to return the files that have a possible match (which can still be done with `fgrep`). In `keg_contain?`, we can check (using ruby) whether the match is a valid one or not. So, `each_unique_file_matching` might return include a file with `/foo/usr/local` but that wouldn't be treated that as a match by `keg_contain?`

Right now, each file found in `each_unique_file_matching` is checked using the output from `strings -t x - file`. Based on some quick glancing at the output from this command on a binary file, it seems that each line is a single word. I don't see any quotes or spaces anything like that. Right now, a match is found using `.include? string` where `string` is the path (e.g. `/usr/local`. Is it enough to use `.start_with? string` instead, assuming that if the line doesn't start with the path we're looking for it is referring to something that shouldn't be relocated or should we just use `.match? /(?<![a-zA-Z0-9])#{string}/` to be safe?

Using `start_with?` would match `/usr/local/foo` but not `/foo/usr/local/foo`. It also wouldn't match `"/usr/local"` or something like that, so if there's a possibility that a string like that would appear it shouldn't be used. I just don't know enough about binary formats and what `strings` does to know. Let me know if that question didn't make any sense.

Rylan12 commented:
Yeah, you're right. I think this is the line in `pkg-config` (for my own reference: see https://github.com/Homebrew/homebrew-core/pull/52820 and https://github.com/Homebrew/homebrew-core/commit/e8a66a6659c2bf54269664a3f986114861604cdc):

```
2e9a8 /usr/local/lib/pkgconfig:/usr/local/share/pkgconfig:/usr/lib/pkgconfig:/usr/local/Homebrew/Library/Homebrew/os/mac/pkgconfig/11.0
```

Which also tells me that we should probably add `:` to the allowable characters in the regex, right?

Either way, seems that using `match?` is the way to go.

> But it does seem it is sort of behaving like `start_with?` currently anyway. The `File.exist?` check is problematic for anything like the `pkg-config` example. I don't actually think that check should really be there.

Yeah, with the `File.exist?` there that does seem to be the case. If `:` separated lists are the only exceptions we could manually split and check each of those. Is that any better? It looks like the check was added by @sjackman intentionally in https://github.com/Homebrew/brew/pull/5366 so I'm hesitant to just remove it.

Rylan12 commented:
Okay, I've made some changes that should bring this new path checking to `keg_contain?` also. A few things to note:

First, both `keg_contain?` and `replace_text_in_files` use `Keg::relocatable_path_regex` (a new method) to create the regex to match the paths. The constant `Keg::RELOCATABLE_PATH_REGEX_PREFIX` can be modified to exclude more characters from both of these simultaneously in the future if needed.

Second, this assumes that any text (from `strings -t x - binary_file`) with colons in it is a list. The list will be split by `:` and each part evaluated separately. I don't think this should have any consequences _unless_ there can be a colon in a path. In that case, there's a small chance that something could be problematic. I don't think this is something we need to worry about.

**Most importantly**: This breaks bottling for `pkg-config` in its current form:

```console
$ brew bottle --verbose --json pkg-config --only-json-tab
==> Determining pkg-config bottle rebuild...
==> Bottling pkg-config--0.29.2_3.big_sur.bottle.1.tar.gz...
/usr/local/opt/gnu-tar/bin/gtar --create --numeric-owner --format pax --owner 0 --group 0 --sort name --pax-option globexthdr.name=/GlobalHead.%n,exthdr.name=%d/PaxHeaders/%f,delete=atime,delete=ctime --file /Users/rylanpolster/Desktop/pkg-config--0.29.2_3.big_sur.bottle.1.tar pkg-config/0.29.2_3
gzip -f pkg-config-bottle.tar
Warning: String '/usr/local/Homebrew' still exists in these files:
/usr/local/Cellar/pkg-config/0.29.2_3/bin/pkg-config
 --> match '/usr/local/lib/pkgconfig:/usr/local/share/pkgconfig:/usr/lib/pkgconfig:/usr/local/Homebrew/Library/Homebrew/os/mac/pkgconfig/11' at offset 0x2e9b4
Error: Bottle contains non-relocatable reference to /usr/local/Homebrew!
```

`pkg-config` contains a reference to the `os/mac/pkgconfig` directory in the Homebrew repository that wasn't caught before. Now, though, it _is_ caught and, because it's a reference to the Homebrew repository, the bottle refuses to build (per https://github.com/Homebrew/brew/pull/10044). As was stated in that PR:

> Refuse to create bottles which have non-relocatable references to HOMEBREW_LIBRARY. This allows us to make all bottles ignore where HOMEBREW_REPOSITORY is (even those that aren't cellar :any). I cannot see any circumstances in which any bottle should link to anything within HOMEBREW_REPOSITORY.

I think that this is the one valid reason where a bottle should be allowed to link to the repository (because we have a designated `pkgconfig` directory in there). Since this is really a one-off, I think it's okay to simply ignore the error if the forbidden path matches `HOMEBREW_REPOSITORY/os/*/pkgconfig`. We may even want to still throw a warning in this situation even though it won't fail.

If I manually ignore the error, the bottle is now correctly recognized as being non-relocatable (i.e. no `cellar` parameter meaning the default must be used). This will allow the `pour_bottle?` block to be removed.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bottle.rb:
```diff
@@ -149,26 +153,34 @@ def keg_contain?(string, keg, ignores, formula_and_runtime_deps_names = nil, arg
           str = io.readline.chomp
           next if ignores.any? { |i| i =~ str }
-          next unless str.include? string
+          next unless str.match? Keg.relocatable_path_regex(string)
```
Is it work caching `Keg.relocatable_path_regex(string)` given it's used multiple times?

Rylan12 replied:
Probably not necessary but can't hurt I guess

MikeMcQuaid commented on Library/Homebrew/keg_relocate.rb:
```diff
@@ -15,6 +17,10 @@ def initialize(**kwargs)
   end
 
+  def self.relocatable_path_regex(path)
+    Regexp.new(RELOCATABLE_PATH_REGEX_PREFIX.source + path)
```
Does this need escaped?

Rylan12 commented:
Ah, I've got to refactor this after #11286. I'll do that later this afternoon. Looks like it will make things easier, actually.

sjackman commented:
You likely noticed I already, but I did in fact flag in https://github.com/Homebrew/brew/pull/10044#issuecomment-747871188 that `pkg-config` contains a reference to `@@HOMEBREW_REPOSITORY@@/Library`.

sjackman commented:
As I said in that comment "`pkg-config` is a bit of a special snow flake", so I'm fine with making an exception for it.

Rylan12 commented:
I added some basic tests for the `Keg::Relocation` class. It's not comprehensive for keg relocation or bottling but it does add tests that check the major changes made in this PR.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bottle.rb:
```diff
@@ -149,26 +153,36 @@ def keg_contain?(string, keg, ignores, formula_and_runtime_deps_names = nil, arg
         until io.eof?
```
I'd be tempted to pull this entire block into a separate function (perhaps in `keg-relocate.rb`?; it seems relatively independent? Np if it's less than I thought.

Rylan12 replied:
I went ahead and extracted it.

I agree, it makes sense to have it separated. It does has a handful of parameters that need to be passed so take a look at if you think it shouldn't really belong in `keg_relocate` I can certainly move it back to `bottle.rb` in a separate function.

carlocab commented on Library/Homebrew/keg_relocate.rb:
```diff
@@ -22,7 +24,7 @@ def freeze
     sig { params(key: Symbol, old_value: T.any(String, Regexp), new_value: String).void }
     def add_replacement_pair(key, old_value, new_value)
-      @replacement_map[key] = [old_value, new_value]
+      @replacement_map[key] = [self.class.path_regex(old_value), new_value]
```
Does this affect the `start_with?` tests in [`relocate_dynamic_linkage`](https://github.com/Homebrew/brew/blob/2b9fd5460e357b6638944f4f89917a669a5af68c/Library/Homebrew/extend/os/mac/keg_relocate.rb#L21)?

Rylan12 replied:
Ugh, yeah it does.

Also, do we still want this behavior when `old_value` is a placeholder in [`prepare_relocation_to_locations`](https://github.com/Homebrew/brew/blob/2b9fd5460e357b6638944f4f89917a669a5af68c/Library/Homebrew/keg_relocate.rb#L94-L102)? I don't think it will cause any problems because the only places a placeholder would be is where the path is already guaranteed to match the regex, but it feels a bit weird.

Rylan12 replied:
Actually, I think we _don't_ want this regex for pathnames because it won't properly replace the placeholders for any existing bottles that have the placeholders in the middle (e.g. `objc-run`).

Rylan12 commented on Library/Homebrew/keg_relocate.rb:
```diff
@@ -70,14 +88,14 @@ def relocate_dynamic_linkage(_relocation)
+    relocation.add_replacement_pair(:library, HOMEBREW_LIBRARY.to_s, LIBRARY_PLACEHOLDER, path: true)
     relocation.add_replacement_pair(:perl,
                                     %r{\A#!(?:/usr/bin/perl\d\.\d+|#{HOMEBREW_PREFIX}/opt/perl/bin/perl)( |$)}o,
                                     "#!#{PERL_PLACEHOLDER}\\1")
```
I didn't add `path: true` here because it looks like it's matching shebangs not paths (cc @Bo98)

Bo98 replied:
That is correct.

I wonder if `path` could be inferred if the first parameter is a `Pathname`. I'm probably overcomplicating it though.

Rylan12 replied:
A tricky thing is that there are some times (e.g. just above this) where a string is passed but should still be treated as a pathname. So, yeah, I think that could be done but we probably need the `path` argument anyway.

I'm leaning toward leaving as is unless you think it makes more sense to automatically detect it.

Rylan12 commented:
Okay, I've updated this to only check paths using the new regex when specifically requested (i.e. only in `prepare_relocation_to_placeholders`).

I also added some more tests for some of the `Keg` methods to test out the placeholders a little bit.

carlocab commented on Library/Homebrew/extend/os/mac/keg_relocate.rb:
```diff
@@ -30,9 +30,9 @@ def relocate_dynamic_linkage(relocation)
         each_install_name_for(file) do |old_name|
-          if old_name.start_with? old_cellar
+          if relocation.start_with_old_value? :cellar, old_name
             new_name = old_name.sub(old_cellar, new_cellar)
-          elsif old_name.start_with? old_prefix
+          elsif relocation.start_with_old_value? :prefix, old_name
             new_name = old_name.sub(old_prefix, new_prefix)
```
This breaks relocation, I'm afraid:
```
❯ otool -L *
libbar.dylib:
        @@HOMEBREW_PREFIX@@/opt/foo/lib/libbar.dylib (compatibility version 0.0.0, current version 0.0.0)
        /usr/local/opt/llvm/lib/libLLVM.dylib (compatibility version 1.0.0, current version 12.0.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.100.5)
libfoo.dylib:
        @@HOMEBREW_PREFIX@@/opt/foo/lib/libfoo.dylib (compatibility version 0.0.0, current version 0.0.0)
        /usr/local/opt/llvm/lib/libLLVM.dylib (compatibility version 1.0.0, current version 12.0.0)
        @loader_path/libbar.dylib (compatibility version 0.0.0, current version 0.0.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.100.5)
❯ otool -l * | rg -A2 LC_RPATH
          cmd LC_RPATH
      cmdsize 40
         path /usr/local/opt/llvm/lib (offset 12)
--
          cmd LC_RPATH
      cmdsize 32
         path @loader_path/ (offset 12)
--
          cmd LC_RPATH
      cmdsize 40
         path /usr/local/opt/llvm/lib (offset 12)
--
          cmd LC_RPATH
      cmdsize 32
         path @loader_path/ (offset 12)
```

carlocab commented on Library/Homebrew/extend/os/mac/keg_relocate.rb:
```diff
@@ -41,9 +41,9 @@ def relocate_dynamic_linkage(relocation)
           each_rpath_for(file) do |old_name|
-            new_name = if old_name.start_with? old_cellar
+            new_name = if relocation.start_with_old_value? :cellar, old_name
               old_name.sub(old_cellar, new_cellar)
-            elsif old_name.start_with? old_prefix
+            elsif relocation.start_with_old_value? :prefix, old_name
```
Relocation was supposed to happen here but didn't:
```
❯ otool -l * | rg -A2 LC_RPATH
          cmd LC_RPATH
      cmdsize 40
         path /usr/local/opt/llvm/lib (offset 12)
--
          cmd LC_RPATH
      cmdsize 32
         path @loader_path/ (offset 12)
--
          cmd LC_RPATH
      cmdsize 40
         path /usr/local/opt/llvm/lib (offset 12)
--
          cmd LC_RPATH
      cmdsize 32
         path @loader_path/ (offset 12)
```

Rylan12 replied:
What's supposed to happen here?

When testing on `master` I get the same output

carlocab replied:
This is the correct output:
```
❯ otool -l * | rg -A2 LC_RPATH
          cmd LC_RPATH
      cmdsize 48
         path @@HOMEBREW_PREFIX@@/opt/llvm/lib (offset 12)
--
          cmd LC_RPATH
      cmdsize 32
         path @loader_path/ (offset 12)
--
          cmd LC_RPATH
      cmdsize 48
         path @@HOMEBREW_PREFIX@@/opt/llvm/lib (offset 12)
--
          cmd LC_RPATH
      cmdsize 32
         path @loader_path/ (offset 12)
```

carlocab reviewed (changes requested):
These new changes break binary relocation, unfortunately.

Rylan12 commented:
Ugh, alright... I'll have another look.

What did you do to test it? Just build a random bottle and check the dylib ids in the bottle archive?

carlocab commented:
I have an actual formula called `foo`. 😄 See previous comment.

Rylan12 commented:
Alright, let's give this another go. I've added another test to catch the issue that caused the rpath relocation to stop working (and fixed the problem).

Rylan12 commented:
Unless there are more comments, I'll plan on merging a little bit later today. I'll try to be around after to make sure there are no major problems.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bottle.rb:
```diff
@@ -31,6 +31,10 @@
 MAXIMUM_STRING_MATCHES = 100
 
+ALLOWABLE_HOMEBREW_REPOSITORY_LINKS = [
+  %r{#{HOMEBREW_LIBRARY}/Homebrew/os/(mac|linux)/pkgconfig},
```
```suggestion
  %r{#{Regexp.escape(HOMEBREW_LIBRARY)}/Homebrew/os/(mac|linux)/pkgconfig},
```

Not sure if this is needed?

Rylan12 replied:
Yeah, doesn't hurt.

MikeMcQuaid commented on Library/Homebrew/extend/os/mac/keg_relocate.rb:
```diff
@@ -131,9 +131,10 @@ def relocated_name_for(old_name, relocation)
     old_cellar, new_cellar = relocation.replacement_pair_for(:cellar)
 
-    if old_name.start_with? old_cellar
+    case old_name
+    when /^#{old_cellar}/
```
What's the difference/preference here?

Rylan12 replied:
Hmm, good question. I think I had it in my mind that it needed to be changed because `start_with?` couldn't be called on a regex but looking at it now, `old_{cellar,prefix}` is the regex not `old_name`, so I think I just got them confused. I'll change it back to the original and see if that works as expected.

MikeMcQuaid commented on Library/Homebrew/keg_relocate.rb:
```diff
@@ -45,6 +48,17 @@ def replace_text(text)
+      when String
+        Regexp.escape(value)
+      when Regexp
+        value.source
```
I assume this is already escaped?

Rylan12 replied:
The assumption is that if a regex is passed, it shouldn't need to be escaped. If I wanted to match `/foo.bar/` and passed it to this function, I would want the resulting regex to be `/(?<![a-zA-Z0-9])foo.bar/` not `/(?<![a-zA-Z0-9])foo\.bar/`.

MikeMcQuaid commented on Library/Homebrew/keg_relocate.rb:
```diff
@@ -45,6 +48,17 @@ def replace_text(text)
     end
+
+    sig { params(value: T.any(String, Regexp)).returns(Regexp) }
+    def self.path_regex(value)
```
```suggestion
    def self.path_to_regex(path)
```

Suggestion here, not binding at all and feel free to disagree.

MikeMcQuaid reviewed (approved):
Some non-blocking comments. This makes sense to me, nice work!

Rylan12 commented:
Okay, this last commit is pretty minimal. Pretty much just style changes from the review. I've retested, though, so I feel good that the rpath relocation still works as expected.

carlocab reviewed (approved):
Also tested on `foo`, works for me. Thanks @Rylan12!
